File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.py
class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        n = len(heights)
        h = []
        def push(x):
            heappush(h, -x)
        def pop():
            return -heappop(h)
        for i in range(1, n):
            if heights[i] > heights[i-1]:
                reqBricks = heights[i] - heights[i-1]
                # Use these many bricks
                push(reqBricks)
                bricks -= reqBricks
                # Dont have enough! -> go back (including this index) 
                # and replace bricks -> ladders
                while ladders > 0 and len(h) and bricks < 0:
                    bricks += pop()
                    ladders -= 1
                if bricks < 0:
                    # we can reach only previous index range(1, n)
                    return i-1

        return n-1



                

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.java
class Solution {
    public int furthestBuilding(int[] heights, int bricks, int ladders) {

        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b-a);

        int n = heights.length;

        for (int i = 1; i < n; i++) {
            if (heights[i] <= heights[i-1]) {
                continue;
            }
            int diff = heights[i] - heights[i-1];
            pq.add(diff);
            bricks -= diff;
            while (ladders > 0 && !pq.isEmpty() && bricks < 0) {
                bricks += pq.poll();
                ladders -= 1;
            } 

            if (bricks < 0) {
                return i - 1;

            }

        }


        return n-1;
    }
}